<!doctype html><html lang=en-us><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=robots content="noodp"><title>How to install BLAST on mac and download the database - Wassim Seifeddine</title><meta name=Description content="Talking about Machine learning, mostly computational biology, bio foundation models and large scale training"><meta property="og:url" content="https://wassimseifeddine.com/posts/blast_p1/"><meta property="og:site_name" content="Wassim Seifeddine"><meta property="og:title" content="How to install BLAST on mac and download the database"><meta property="og:description" content="BLAST (Basic Local Alignment Search Tool) is a widely used tool for comparing nucleotide or protein sequences against databases. If you’re working on macOS and need to use any of its tools, this guide will help you install BLAST, download necessary databases, and run BLAST searches efficiently.
Installing Manually Installing BLAST manually gives you control over the installation process and allows you to install the latest version directly from NCBI."><meta property="og:locale" content="en_us"><meta property="og:type" content="article"><meta property="article:section" content="posts"><meta property="article:published_time" content="2024-04-24T14:05:13+02:00"><meta property="article:modified_time" content="2024-04-24T14:05:13+02:00"><meta name=twitter:card content="summary"><meta name=twitter:title content="How to install BLAST on mac and download the database"><meta name=twitter:description content="BLAST (Basic Local Alignment Search Tool) is a widely used tool for comparing nucleotide or protein sequences against databases. If you’re working on macOS and need to use any of its tools, this guide will help you install BLAST, download necessary databases, and run BLAST searches efficiently.
Installing Manually Installing BLAST manually gives you control over the installation process and allows you to install the latest version directly from NCBI."><meta name=application-name content="Wassim Seifeddine"><meta name=apple-mobile-web-app-title content="Wassim Seifeddine"><meta name=theme-color content="#ffffff"><meta name=msapplication-TileColor content="#da532c"><link rel="shortcut icon" type=image/x-icon href=/favicon.ico><link rel=icon type=image/png sizes=32x32 href=/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/favicon-16x16.png><link rel=apple-touch-icon sizes=180x180 href=/apple-touch-icon.png><link rel=mask-icon href=/safari-pinned-tab.svg color=#5bbad5><link rel=manifest href=/site.webmanifest><link rel=canonical href=https://wassimseifeddine.com/posts/blast_p1/><link rel=prev href=https://wassimseifeddine.com/posts/intro/><link rel=next href=https://wassimseifeddine.com/posts/torchtitan/><link rel=stylesheet href=/css/style.min.css><link rel=preload href=https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.1.1/css/all.min.css as=style onload='this.onload=null,this.rel="stylesheet"'><noscript><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.1.1/css/all.min.css></noscript><link rel=preload href=https://cdn.jsdelivr.net/npm/animate.css@4.1.1/animate.min.css as=style onload='this.onload=null,this.rel="stylesheet"'><noscript><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/animate.css@4.1.1/animate.min.css></noscript><script type=application/ld+json>{"@context":"http://schema.org","@type":"BlogPosting","headline":"How to install BLAST on mac and download the database","inLanguage":"en-us","mainEntityOfPage":{"@type":"WebPage","@id":"https:\/\/wassimseifeddine.com\/posts\/blast_p1\/"},"genre":"posts","wordcount":1111,"url":"https:\/\/wassimseifeddine.com\/posts\/blast_p1\/","datePublished":"2024-04-24T14:05:13+02:00","dateModified":"2024-04-24T14:05:13+02:00","publisher":{"@type":"Organization","name":""},"author":{"@type":"Person","name":"Author"},"description":""}</script></head><body data-header-desktop=fixed data-header-mobile=auto><script type=text/javascript>(window.localStorage&&localStorage.getItem("theme")?localStorage.getItem("theme")==="dark":"light"==="auto"?window.matchMedia("(prefers-color-scheme: dark)").matches:"light"==="dark")&&document.body.setAttribute("theme","dark")</script><div id=mask></div><div class=wrapper><header class=desktop id=header-desktop><div class=header-wrapper><div class=header-title><a href=/ title="Wassim Seifeddine">Wassim Seifeddine</a></div><div class=menu><div class=menu-inner><a class=menu-item href=/about/>About </a><a class=menu-item href=/portfolio/>Portfolio </a><a class=menu-item href=/posts/>Posts </a><a class=menu-item href=/conferences/>Conferences </a><a class=menu-item href="https://scholar.google.com/citations?user=x6EK09kAAAAJ&amp;hl=en" rel="noopener noreffer" target=_blank>Publications </a><a class=menu-item href=/seifeddine_wassim_cv.pdf>Resume </a><span class="menu-item delimiter"></span><a href=javascript:void(0); class="menu-item theme-switch" title="Switch Theme">
<i class="fas fa-adjust fa-fw" aria-hidden=true></i></a></div></div></div></header><header class=mobile id=header-mobile><div class=header-container><div class=header-wrapper><div class=header-title><a href=/ title="Wassim Seifeddine">Wassim Seifeddine</a></div><div class=menu-toggle id=menu-toggle-mobile><span></span><span></span><span></span></div></div><div class=menu id=menu-mobile><a class=menu-item href=/about/ title>About</a><a class=menu-item href=/portfolio/ title>Portfolio</a><a class=menu-item href=/posts/ title>Posts</a><a class=menu-item href=/conferences/ title>Conferences</a><a class=menu-item href="https://scholar.google.com/citations?user=x6EK09kAAAAJ&amp;hl=en" title rel="noopener noreffer" target=_blank>Publications</a><a class=menu-item href=/seifeddine_wassim_cv.pdf title>Resume</a><a href=javascript:void(0); class="menu-item theme-switch" title="Switch Theme">
<i class="fas fa-adjust fa-fw" aria-hidden=true></i></a></div></div></header><main class=main><div class=container><div class=toc id=toc-auto><h2 class=toc-title>Contents</h2><div class=toc-content id=toc-content-auto></div></div><article class="page single"><h1 class="single-title animate__animated animate__flipInX">How to install BLAST on mac and download the database</h1><div class=post-meta><div class=post-meta-line><span class=post-author><a href=/ title=Author rel=author class=author><i class="fas fa-user-circle fa-fw" aria-hidden=true></i>Author</a></span></div><div class=post-meta-line><i class="far fa-calendar-alt fa-fw" aria-hidden=true></i>&nbsp;<time datetime=2024-04-24>2024-04-24</time>&nbsp;<i class="fas fa-pencil-alt fa-fw" aria-hidden=true></i>&nbsp;1111 words&nbsp;
<i class="far fa-clock fa-fw" aria-hidden=true></i>&nbsp;6 minutes&nbsp;</div></div><div class="details toc" id=toc-static data-kept><div class="details-summary toc-title"><span>Contents</span>
<span><i class="details-icon fas fa-angle-right" aria-hidden=true></i></span></div><div class="details-content toc-content" id=toc-content-static><nav id=TableOfContents><ul><li><a href=#installing>Installing</a><ul><li><a href=#manually>Manually</a></li><li><a href=#via-homebrew>Via Homebrew</a></li></ul></li><li><a href=#downloading-the-database>Downloading the Database</a></li><li><a href=#running-blast>Running BLAST</a></li><li><a href=#conclusion>Conclusion</a></li><li><a href=#references>References</a></li></ul></nav></div></div><div class=content id=content><p>BLAST (Basic Local Alignment Search Tool) is a widely used tool for comparing nucleotide or protein sequences against databases. If you&rsquo;re working on macOS and need to use any of its tools, this guide will help you install BLAST, download necessary databases, and run BLAST searches <em>efficiently</em>.</p><h2 id=installing>Installing</h2><h3 id=manually>Manually</h3><p>Installing BLAST manually gives you control over the installation process and allows you to install the latest version directly from NCBI.</p><ol><li><p><strong>Download the BLAST Package:</strong></p><ul><li>Visit the NCBI BLAST download page: <a href=https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ target=_blank rel="noopener noreffer">NCBI BLAST+ Download</a></li><li>Download the latest macOS binary package<ol><li>For intel based mac typically the file name should be like <code>ncbi-blast-x.xx.x+-x64-macosx.tar.gz</code>.</li><li>For apple silicone based mac typically the file name should be like <code>ncbi-blast-x.xx.x+-aarch64-macosx.tar.gz </code>.</li></ol></li></ul></li></ol><p>with <code>x.xx.x</code> as the version. As the time of writing this guide the latest version is <code>2.16.0</code>.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl><span class=nb>cd</span> ~/Downloads
</span></span><span class=line><span class=cl>curl -O https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.13.0+-x64-macosx.tar.gz
</span></span></code></pre></div><p><strong>Note:</strong> Replace the URL with correct one for your platform.</p><ol start=2><li><strong>Extract the Package:</strong></li></ol><p>Open Terminal and navigate to the download directory:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl><span class=nb>cd</span> ~/Downloads
</span></span><span class=line><span class=cl>tar -zxvf ncbi-blast-2.16.0+-x64-macosx.tar.gz
</span></span></code></pre></div><ol start=3><li><strong>You can now use BLAST binaries in the <code>ncbi-blast-2.16.0+/bin</code> directory.</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl><span class=nb>cd</span> ncbi-blast-2.16.0+/bin
</span></span><span class=line><span class=cl>./blastn -version
</span></span></code></pre></div><p>If everything is working correctly, you should see the BLAST version information.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>blastn: 2.16.0+
</span></span><span class=line><span class=cl>Package: blast 2.16.0, build Jun <span class=m>25</span> <span class=m>2024</span> 11:53:51
</span></span></code></pre></div><ol start=4><li><strong>Making BLAST available anywhere</strong></li></ol><p>Right now if you want to use BLAST, you have to be in the <code>ncbi-blast-2.16.0+/bin</code> directory. To make it accessible anywhere, there are two solutions. Either move the binaries into where your shell ( terminal ) searches for binaries, or tell it to look somewhere else (where BLAST is installed).
I normally prefer the second optione because</p><ol><li>It gives me more flexibility on where I can install and manage third party tools</li><li>I can have multiple version of BLAST and switch between them in one command</li></ol><p>The way to accomplish this is by adding the directory where BLAST is installed to the <code>PATH</code> environment variable in macOS. The <code>PATH</code> variable simply contains a list of directories that macos uses to seacrh for binaries.</p><p>To add a directory into this variable, simply locate where BLAST is installed, for my setup, I keep such tools in <code>~/third_party/blast/&lt;version>/bin</code> so for version <code>2.16.0</code>, the full directory path would be <code>~/third_party/blast/2.16.0/bin</code></p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl><span class=nb>export</span> <span class=nv>PATH</span><span class=o>=</span><span class=s2>&#34;</span><span class=nv>$HOME</span><span class=s2>/third_party/blast/2.16.0/bin:</span><span class=nv>$PATH</span><span class=s2>&#34;</span>
</span></span></code></pre></div><p><strong>Note</strong>: This applies only to your current SHELL session, if you open a new terminal, you&rsquo;ll have to execute this command again to make BLAST accessbile for this session.</p><p>To make it persistant across sessions, you&rsquo;ll have to add the command to <code>~/.bashrc</code> or <code>~/.zshrc</code>.</p><ol start=6><li><strong>Verify Installation:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>blastn -version
</span></span></code></pre></div><p>You should see the BLAST version information.</p><h3 id=via-homebrew>Via Homebrew</h3><p>Homebrew simplifies software installation on macOS.</p><ol><li><strong>Install Homebrew (if not already installed):</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>/bin/bash -c <span class=s2>&#34;</span><span class=k>$(</span>curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh<span class=k>)</span><span class=s2>&#34;</span>
</span></span></code></pre></div><ol start=2><li><strong>Install BLAST via Homebrew:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>brew install blast
</span></span></code></pre></div><ol start=3><li><strong>Verify Installation:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>blastn -version
</span></span></code></pre></div><h2 id=downloading-the-database>Downloading the Database</h2><p>BLAST requires databases to compare your query sequences against. Their database is available <a href=https://ftp.ncbi.nlm.nih.gov/blast/db/ target=_blank rel="noopener noreffer">here</a>.</p><ol><li><p><strong>Choose a Database:</strong></p><ul><li>Common databases:<ul><li><code>nt</code>: Nucleotide collection. It is used for nucleotide sequence searches, so if you have a DNA or RNA sequence and want to find similar nucleotide sequences, you would use the nt database.</li><li><code>nr</code>: Non-redundant protein sequences. It is used for protein sequence searches, so if you have a protein sequence and want to find similar proteins, you would use the nr database.</li><li>Specific organism databases</li></ul></li></ul></li><li><p><strong>Create a Directory for Databases:</strong></p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>mkdir -p ~/blastdb
</span></span></code></pre></div><ol start=3><li><strong>Download Databases Using <code>update_blastdb.pl</code>:</strong></li></ol><p>You can download the database manually from the link above then extract them. However, BLAST provides a script to automatically do it.
The script is called <code>update_blastdb.pl</code> and is included in the <code>ncbi-blast-2.16.0+/bin</code> directory.
The <code>update_blastdb.pl</code> script is included with BLAST. Since you made the full <code>bin</code> directory available in your <code>PATH</code>, you can also run this script from anywhere.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>update_blastdb.pl --decompress nt  ~/blastdb <span class=c1># for nt database</span>
</span></span><span class=line><span class=cl>update_blastdb.pl --decompress nr  ~/blastdb <span class=c1># for nr database</span>
</span></span></code></pre></div><p><em>Note:</em> Replace <code>nt</code> with the database of your choice.
The problem is that the database is huge and you probably shouldn&rsquo;t download all of it in your mac if you have another option. Normally I have a small database locally that I use to test scripts and when i&rsquo;m confident with the script, I run it on the server where the full database is available.</p><ol start=3><li><strong>Download a partial database</strong>
You can download only one part of the database by specifying the part you want to download. For example, to download the first file of the <code>nt</code> or <code>nr</code> database, you can run the following command:</li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>update_blastdb.pl --decompress nt.00  ~/blastdb <span class=c1># for nt database</span>
</span></span></code></pre></div><p>or using curl</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>curl -O https://ftp.ncbi.nlm.nih.gov/blast/db/nr.00.tar.gz
</span></span><span class=line><span class=cl>tar -zxvf nr.00.tar.gz
</span></span></code></pre></div><p><strong>Note</strong>: This file is 33GB in size before uncompressing, so make sure you have enough space on your mac.</p><ol start=4><li><strong>Set the <code>BLASTDB</code> Environment Variable:</strong>
Like the terminal needs to know where to search for binaries, BLAST needs to know where to search for databases. You can set the <code>BLASTDB</code> environment variable to the directory where you downloaded the databases.</li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl><span class=nb>export</span> <span class=nv>BLASTDB</span><span class=o>=</span>~/blastdb
</span></span></code></pre></div><p>This also applies only to your current SHELL session, if you open a new terminal, you&rsquo;ll have to execute this command again to make BLASTDB accessbile to BLAST. To make it persistant across session, you&rsquo;ll have to add the command to <code>~/.bashrc</code> or <code>~/.zshrc</code>.</p><ol start=5><li><strong>Verify Database Files:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>ls ~/blastdb
</span></span></code></pre></div><h2 id=running-blast>Running BLAST</h2><p>With BLAST installed and databases downloaded, you can perform searches.</p><ol><li><strong>Prepare Your Query Sequence File:</strong></li></ol><p>Save your sequence in FASTA format (e.g., <code>query.fasta</code>).
You can try the below sequence as a test.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>&gt;Seq1
</span></span><span class=line><span class=cl>ATGGGTAAGGAGGACAAGACTCACCTTAACGT
</span></span><span class=line><span class=cl>CGTCGTCATCGGCCACGTCGACTCTGGCAAGT
</span></span><span class=line><span class=cl>CGACCACTGTAAGTACAACCAACAGCGGGTTG
</span></span><span class=line><span class=cl>CTTATCTGCACTCGGAATCCGCCAAACCTGGC
</span></span><span class=line><span class=cl>AGGGTATCACCAAAACATCTTGCTAACTTTTG
</span></span><span class=line><span class=cl>ACAGACCGGTCACTTGATCTACCAGTGCGGTG
</span></span><span class=line><span class=cl>GTATCGACAAGCGAACCATCGAGAAGTTCGAG
</span></span><span class=line><span class=cl>AAGGTTAGTCAATATCCCTTCGATTACGCGCG
</span></span><span class=line><span class=cl>CTCCCATCGATTCCCACGATTCGCTCCCTCAC
</span></span><span class=line><span class=cl>TCGAAACACATCCATTACCCCGCTCGAGTCCG
</span></span><span class=line><span class=cl>AAAATTTTGCGGTGCGACCGTGATTTTTTCTG
</span></span><span class=line><span class=cl>GTGGGGTATCTTACCCCGCCACTCGAGTCACG
</span></span><span class=line><span class=cl>GATGCGCTTGCCCTGTTCCCACAAAACCTTAC
</span></span><span class=line><span class=cl>CACCCTGTCGCGCACTACATGTCTTGCAGTCA
</span></span><span class=line><span class=cl>CTAACCACTGGACAATAGGAAGCCGCCGAGCT
</span></span><span class=line><span class=cl>CGGAAAGGGTTCCTTCAAGTACGCCTGGGTTC
</span></span><span class=line><span class=cl>TTGACAAGCTCAAAGCCGAGCGTGAGCGTGGT
</span></span><span class=line><span class=cl>ATCACCATTGATATCGCTCTCTGGAAGTTCGA
</span></span><span class=line><span class=cl>GACTCCTCGCTACTATGTCACCGTCATTGGTA
</span></span><span class=line><span class=cl>TGTTGTCACCGTCTCACACTATCATGTATTCA
</span></span><span class=line><span class=cl>TCATGCTAACATCTCTCTCAGATGCCCCCGGT
</span></span><span class=line><span class=cl>CATCGTGATTTCATCAAGAACATGATC 
</span></span></code></pre></div><ol start=2><li><strong>Run a BLAST Search:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>blastn -query query.fasta -db nt -out results.txt
</span></span></code></pre></div><ul><li><code>blastn</code>: Nucleotide-nucleotide BLAST</li><li><code>-query</code>: Input file</li><li><code>-db</code>: Database name</li><li><code>-out</code>: Output file</li></ul><ol start=3><li><strong>Customize Search Parameters (Optional):</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>blastn -query query.fasta -db nt -out results.txt -evalue 1e-5 -outfmt <span class=m>6</span> -max_target_seqs <span class=m>10</span>
</span></span></code></pre></div><ul><li><code>-evalue</code>: Expectation value threshold</li><li><code>-outfmt 6</code>: Tabular output format</li><li><code>-max_target_seqs</code>: Maximum number of aligned sequences to keep</li></ul><p><strong>Note</strong>: if you are getting any error related to the database, make sure that:</p><ol><li><code>BLASTDB</code> environment variable is set correctly.</li><li>The database is downloaded and decompressed correctly.</li><li>The database is in the correct directory.</li><li>You can run <code>blastdbcmd -db nt -info</code> to check if the database is correctly installed.<ul><li>Make sure you are using the correct database name in the command.</li></ul></li><li>Make sure you are using the correct command.<ul><li><code>blastn</code> for nucleotide-nucleotide BLAST</li><li><code>blastp</code> for protein-protein BLAST</li></ul></li></ol><br><ol start=4><li><strong>View Results:</strong></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>less results.txt
</span></span></code></pre></div><h2 id=conclusion>Conclusion</h2><p>Installing BLAST on macOS is straightforward, whether you choose a manual installation or use Homebrew. With BLAST set up and databases downloaded, you&rsquo;re ready to perform powerful sequence analyses directly from your Mac. In the next guide, I&rsquo;ll go over how to interpret the results from BLAST.</p><h2 id=references>References</h2><ul><li><a href=https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ target=_blank rel="noopener noreffer">NCBI BLAST+ Download</a></li><li><a href=https://www.ncbi.nlm.nih.gov/books/NBK279690/ target=_blank rel="noopener noreffer">BLAST Command Line Applications User Manual</a></li><li><a href=https://brew.sh/ target=_blank rel="noopener noreffer">Homebrew</a></li><li><a href=https://ftp.ncbi.nlm.nih.gov/blast/db/ target=_blank rel="noopener noreffer">NCBI BLAST Databases</a></li></ul></div><div class=post-footer id=post-footer><div class=post-info><div class=post-info-line><div class=post-info-mod><span>Updated on 2024-04-24</span></div></div><div class=post-info-line><div class=post-info-md></div><div class=post-info-share><span><a href=javascript:void(0); title="Share on Twitter" data-sharer=twitter data-url=https://wassimseifeddine.com/posts/blast_p1/ data-title="How to install BLAST on mac and download the database"><i class="fab fa-twitter fa-fw" aria-hidden=true></i></a><a href=javascript:void(0); title="Share on Facebook" data-sharer=facebook data-url=https://wassimseifeddine.com/posts/blast_p1/><i class="fab fa-facebook-square fa-fw" aria-hidden=true></i></a><a href=javascript:void(0); title="Share on Hacker News" data-sharer=hackernews data-url=https://wassimseifeddine.com/posts/blast_p1/ data-title="How to install BLAST on mac and download the database"><i class="fab fa-hacker-news fa-fw" aria-hidden=true></i></a><a href=javascript:void(0); title="Share on Line" data-sharer=line data-url=https://wassimseifeddine.com/posts/blast_p1/ data-title="How to install BLAST on mac and download the database"><i data-svg-src=https://cdn.jsdelivr.net/npm/simple-icons@7.3.0/icons/line.svg aria-hidden=true></i></a><a href=javascript:void(0); title="Share on 微博" data-sharer=weibo data-url=https://wassimseifeddine.com/posts/blast_p1/ data-title="How to install BLAST on mac and download the database"><i class="fab fa-weibo fa-fw" aria-hidden=true></i></a></span></div></div></div><div class=post-info-more><section class=post-tags></section><section><span><a href=javascript:void(0); onclick=window.history.back()>Back</a></span>&nbsp;|&nbsp;<span><a href=/>Home</a></span></section></div><div class=post-nav><a href=/posts/intro/ class=prev rel=prev title=Intro><i class="fas fa-angle-left fa-fw" aria-hidden=true></i>Intro</a>
<a href=/posts/torchtitan/ class=next rel=next title="Torchtitan: A PyTorch Library for Parallelism Techniques explained">Torchtitan: A PyTorch Library for Parallelism Techniques explained<i class="fas fa-angle-right fa-fw" aria-hidden=true></i></a></div></div></article></div></main><footer class=footer><div class=footer-container><div class=footer-line itemscope itemtype=http://schema.org/CreativeWork></div></div></footer></div><div id=fixed-buttons><a href=# id=back-to-top class=fixed-button title="Back to Top"><i class="fas fa-arrow-up fa-fw" aria-hidden=true></i>
</a><a href=# id=view-comments class=fixed-button title="View Comments"><i class="fas fa-comment fa-fw" aria-hidden=true></i></a></div><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css><script type=text/javascript src=https://cdn.jsdelivr.net/npm/lazysizes@5.3.2/lazysizes.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/clipboard@2.0.11/dist/clipboard.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/sharer.js@0.5.1/sharer.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/auto-render.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/copy-tex.min.js></script><script type=text/javascript src=https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/mhchem.min.js></script><script type=text/javascript>window.config={code:{copyTitle:"Copy to clipboard",maxShownLines:50},comment:{},math:{delimiters:[{display:!0,left:"$$",right:"$$"},{display:!0,left:"\\[",right:"\\]"},{display:!0,left:"\\begin{equation}",right:"\\end{equation}"},{display:!0,left:"\\begin{equation*}",right:"\\end{equation*}"},{display:!0,left:"\\begin{align}",right:"\\end{align}"},{display:!0,left:"\\begin{align*}",right:"\\end{align*}"},{display:!0,left:"\\begin{alignat}",right:"\\end{alignat}"},{display:!0,left:"\\begin{alignat*}",right:"\\end{alignat*}"},{display:!0,left:"\\begin{gather}",right:"\\end{gather}"},{display:!0,left:"\\begin{CD}",right:"\\end{CD}"},{display:!1,left:"$",right:"$"},{display:!1,left:"\\(",right:"\\)"}],strict:!1}}</script><script type=text/javascript src=/js/theme.min.js></script><script type=text/javascript>window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}gtag("js",new Date),gtag("config","G-JKG8HGHF8S",{anonymize_ip:!0})</script><script type=text/javascript src="https://www.googletagmanager.com/gtag/js?id=G-JKG8HGHF8S" async></script></body></html>